designs/2024-ignore-file-directive/README.md

```diff
@@ -0,0 +1,121 @@
+
+For any _source file_ (the root of an AST), if it contains a leading `/* eslint-ignore-file */` directive, skip parsing it. More concretely, the following pseudocode algorithm explains how to determine whether a file should be ignored or processed:
+
+```
+for line of source file:
+  if line (is not zero or more whitespace) and (does not start with zero or more whitespace + "/"):
+    parse file
+  if line is "/* eslint-ignore-file */":
+    skip file parsing
+```
+
+A more sophisticated algorithm may be chosen at a later point in the lifecycle of this RFC.
+
+In practice, the imaginary
+
+```ts
+export declare function parseFile(filePath: string): AST;
+```
+
+needs to be replaced with:
+
+```ts
+export function parseFile(filePath: string): AST | null {
+  const fileContent = streamFileContent(filePath);
+
+  if (shouldIgnore(fileContent)) return null;
+  return parseFileImpl(fileContent.collectToString());
+}
+
+declare function parseFileImpl(fileContent: string): AST;
+```
+
+How the rest of the system deals with `null` results from parsing a source file is not in the scope of this RFC.
+
+If an `/* eslint-ignore-file */` directive is encountered while parsing a file, a non-fatal error shall be raised, and the directive shall have no further effect.
+
+## Documentation
+
+TBD. The primary audience is a library author. The point **Drawbacks > Bardwards Compatibility** shall be highlighted.
+
+## Drawbacks
+
+1. Backwards Compatibility
+
+Library authors will need to prefix their generated code with both `/* eslint-ignore-file */` and `/* eslint-disable */` directives in order to not break existing projects that have not upgraded their ESLint core to a version that supports the `/* eslint-ignore-file */` directive. Otherwise, updating libraries without updating ESLint accordingly will lead to possible linging errors on generated source files that were previously muted.
+
+2. Primitive Parser
+
+The algorithm described in the **Detailed Design** section is very limited because it does not parse source code. Instead, it directly scans the lines' content. A file containing the following source code is skipped, even though it seems like it shouldn't:
+
+```ts
+
+/* some comment */ statement;
+
+/* eslint-ignore-file */
+```
+
+This is an intentional middleground between keeping the analysis as simple (and fast) as reasonable and correctness, which would require parsing read comments to some degree.
+
+3. `.eslintignore` No Longer Authoritative
+
+By extending how files can be ignored, it is no longer purely `.eslintignore` files that dictate which files are parsed or not.
+
+## Backwards Compatibility Analysis
+
+Other than what is highlighted in **Drawbacks > Bardwards Compatibility**, there are no backwards compatibility issues.
+
+## Alternatives
+
+1. As highlighted in **Motivation**, everything described in this RPC is functionally already possible with `.eslintignore` files.
```

Comment:
If this is already possible using an ignore file, then why would we add another way to do the same thing?

Reply:
Now that this is no longer possible with a local ignore file and strict unused-directive check is on by default I'd like to revisit this, should I open an new proposal?

Reply:
It's possible with `ignorePatterns` in a config, which is a replacement for `.eslintignore`. Why isn't that sufficient?

Reply:
It's hard to locate and modify the top-level eslint.config.js programmatically for a tool that generates files, but easy to add a directive to the generated file or stick an .eslintignore file next to the files.

Currently tools like TanStack Router generate files in a codebase with ignore directives but eslint 9 makes these less useful because they may trigger unused ignore directive warnings. It'd be useful to 0reserve a way to ignore a file from the file itself that doesn't trigger these unused disable warnings which are now enabled by default.